ffi/examples/read-table/expression.h

```diff
@@ -0,0 +1,713 @@
+    .visit_string_literal = visit_expr_string_literal,
+    .visit_struct_literal = visit_expr_struct_literal,
+    .visit_array_literal = visit_expr_array_literal,
+    .visit_and = visit_expr_and,
+    .visit_or = visit_expr_or,
+    .visit_not = visit_expr_not,
+    .visit_is_null = visit_expr_is_null,
+    .visit_lt = visit_expr_lt,
+    .visit_le = visit_expr_le,
+    .visit_gt = visit_expr_gt,
+    .visit_ge = visit_expr_ge,
+    .visit_eq = visit_expr_eq,
+    .visit_ne = visit_expr_ne,
+    .visit_distinct = visit_expr_distinct,
+    .visit_in = visit_expr_in,
+    .visit_not_in = visit_expr_not_in,
+    .visit_add = visit_expr_add,
+    .visit_minus = visit_expr_minus,
+    .visit_multiply = visit_expr_multiply,
+    .visit_divide = visit_expr_divide,
+    .visit_column = visit_expr_column,
+    .visit_struct_expr = visit_expr_struct_expr,
+  };
+  uintptr_t top_level_id = visit_expression(&predicate, &visitor);
+  ExpressionItemList top_level_expr = data.lists[top_level_id];
+  free(data.lists);
+  return top_level_expr;
+}
+
+void free_expression_list(ExpressionItemList list);
+void free_expression_item(ExpressionItem ref)
+{
+  switch (ref.type) {
+    case BinOp: {
+      struct BinOp* op = ref.ref;
+      free_expression_list(op->exprs);
+      free(op);
+      break;
+    }
+    case Variadic: {
+      struct Variadic* var = ref.ref;
+      free_expression_list(var->exprs);
+      free(var);
+      break;
+    };
+    case Literal: {
+      struct Literal* lit = ref.ref;
+      switch (lit->type) {
+        case Struct: {
+          struct Struct* struct_data = &lit->value.struct_data;
+          free_expression_list(struct_data->values);
+          free_expression_list(struct_data->fields);
+          break;
+        }
+        case Array: {
+          struct ArrayData* array = &lit->value.array_data;
+          free_expression_list(array->exprs);
+          break;
+        }
+        case String: {
+          free(lit->value.string_data);
+          break;
+        }
+        case Binary: {
+          struct BinaryData* binary = &lit->value.binary;
+          free(binary->buf);
+          break;
+        }
+        case Integer:
+        case Long:
+        case Short:
+        case Byte:
+        case Float:
+        case Double:
+        case Boolean:
+        case Timestamp:
+        case TimestampNtz:
+        case Date:
+        case Decimal:
+        case Null:
+          break;
+      }
+      free(lit);
+      break;
+    };
+    case Unary: {
+      struct Unary* unary = ref.ref;
+      free_expression_list(unary->sub_expr);
+      free(unary);
+      break;
+    }
+    case Column: {
+      free((void*)ref.ref);
+      break;
+    }
+  }
+}
+
+void free_expression_list(ExpressionItemList list)
+{
+  for (size_t i = 0; i < list.len; i++) {
+    free_expression_item(list.list[i]);
+  }
+  free(list.list);
+}
+
+/*************************************************************
```

Comment:
Suggest putting the printing stuff in its own file (`.h` or `.h`+`.c` as you prefer). This is a big file already :)